pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [connect timed out on COMMIT]

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://mesolite.cluster.engflow.com/invocations/default/62dc5fcd-2c27-4ed2-841e-0f66dde4724a?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvY2NsL3Rlc3RjY2wvd29ya2xvYWQvc2NoZW1hY2hhbmdlOnNjaGVtYWNoYW5nZV90ZXN0) on master @ [614d7b459f0db5d15c8f37d7027e0d45145c017e](https://github.com/cockroachdb/cockroach/commits/614d7b459f0db5d15c8f37d7027e0d45145c017e):

Fatal error:

```
panic: connection has timed out; sessionID=18282a3bc7f0e0f10000000000000001 activeQuery={String:COMMIT TRANSACTION Valid:true}

```
Stack: 

```
goroutine 30744 [running]:
github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWatchDog).watchLoop(0xc00c2f6540, {0x8241620, 0xc00b630be0})
	pkg/workload/schemachange/watch_dog.go:113 +0x2b4
created by github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWatchDog).Start in goroutine 13770
	pkg/workload/schemachange/watch_dog.go:131 +0x111
```
<details><summary>Log preceding fatal error</summary>
<p>

```
  },
  "COMMIT"
 ],
 "expectedExecErrors": "",
 "expectedCommitErrors": "",
 "message": "TXN RETRY ERROR; ERROR: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE - failed preemptive refresh due to conflicting locks on /NamespaceTable/30/1/104/110/\"view_w1_111\"/4/1 [reason=wait_policy] - conflicting txn: meta={id=a65adfb6 key=/Table/3/1/127/2/1 iso=Serializable pri=0.00933600 epo=0 ts=1740688045.883786798,1 min=1740687697.415994048,0 seq=2}): \"sql txn\" meta={id=2e0826b2 key=/Table/119/6/\"E\\xb0\\x84\\xcd_\\x7fE\\xf8\\xbe\\xa4mdL0*Z\"/\"B\\xfd\\x0e\\x85g\\xc5\\bG\\x03\\x0f\\x00\\xc0\\b\\xb3}+\\x81\\x8a\\x8c\\xbf\\xf1np\\xed\\xa9\\xe98?\\xf7\\xa3\\xf3\\\"'\\x15\\xc8?\\xf8\\xed\\xad{E\\xe3!\\xfd\\x0e\\x94{\\x13\\xfa\\x1f\\x80\\x01\\x88\" iso=Serializable pri=0.00393912 epo=0 ts=1740688054.162541464,2 min=1740688036.136629934,0 seq=26} lock=true stat=PENDING rts=1740688036.136629934,0 wto=false gul=1740688036.636629934,0 (SQLSTATE 40001)"
}
{
 "workerId": 1,
 "clientTimestamp": "20:27:37.406438",
 "ops": [
  "BEGIN",
  {
   "sql": "ALTER TABLE schema_w1_2.table_w1_141 ALTER COLUMN \"IrrelevantColumnName\" DROP NOT NULL",
   "expectedExecErr": "42P01"
  }
 ],
 "expectedExecErrors": "42P01",
 "expectedCommitErrors": "",
 "message": "ROLLBACK; Ignoring declarative schema changer not implemented error.: ERROR: *tree.AlterTable not implemented in the new schema changer (SQLSTATE XXUUU)"
}
{
 "workerId": 1,
 "clientTimestamp": "20:27:57.967428",
 "ops": [
  "BEGIN",
  {
   "sql": "SELECT t1.col47_w0_51 AS col0, t0.col_w1_9_w1_35 AS col1, t1.col47_w0_48 AS col2, t1.col47_w0_57 AS col3, t0.col9_w1_19 AS col4 FROM schema_w1_2.table_w1_9 AS t0 ,public.table_w0_47 AS t1 FETCH FIRST 1 ROWS ONLY",
   "potentialExecErr": "53100,53200,57014"
  },
  "COMMIT"
 ],
 "expectedExecErrors": "",
 "expectedCommitErrors": "",
 "message": ""
}
{
 "workerId": 1,
 "clientTimestamp": "20:28:19.113957",
 "ops": [
  "BEGIN",
  {
   "sql": "ALTER TABLE public.table_w0_1 RENAME TO public.table_w1_142"
  },
  "COMMIT"
 ],
 "expectedExecErrors": "",
 "expectedCommitErrors": "",
 "message": ""
}
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #141813 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot T-sql-foundations branch-release-23.2]
- #141696 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot P-2 T-sql-foundations branch-release-23.2.21-rc]
- #141152 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [column &#34;crdb_internal_column_3_name_placeholder&#34; does not exist] [C-test-failure O-robot P-2 T-sql-foundations branch-release-25.1.0-rc]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-48163

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18953348?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18953348?buildTab=artifacts#/) on master @ [281a7294cf1360446177ceed1e25e141d19e0c6e](https://github.com/cockroachdb/cockroach/commits/281a7294cf1360446177ceed1e25e141d19e0c6e):


```
  {
   "sql": "CREATE TABLE public.table_w1_124 (col124_w1_125 REGTYPE NOT NULL, col124_w1_126 BIT(14) NOT NULL, col124_w1_127 STRING COLLATE de_DE, col124_w1_128 REGCLASS NULL, col124_w1_129 TIMETZ NOT NULL, col124_w1_130 INT2 NOT NULL, col124_w1_131 INT4, col124_w1_132 UUID NOT NULL, col124_w1_133 JSONB, col124_w1_134 OID, col124_w1_135 TIME NOT NULL, col124_w1_136 DATE NOT NULL, col124_w1_137 INT2 NOT NULL, col124_w1_138 TSVECTOR NULL, col124_w1_139 BOX2D, col124_w1_140 STRING NOT NULL, col124_w1_141 \"char\" NOT NULL, col124_w1_142 INT8 AS ((col124_w1_137 + col124_w1_131) + col124_w1_130) VIRTUAL, PRIMARY KEY (col124_w1_126, col124_w1_136 ASC, col124_w1_125 DESC), UNIQUE (lower(CAST(col124_w1_135 AS STRING))) STORING (col124_w1_127, col124_w1_130, col124_w1_131, col124_w1_132), INDEX (col124_w1_135, col124_w1_130 ASC))"
  },
  {
   "sql": "ALTER TABLE public.table_w1_104 RENAME TO public.table_w1_143"
  },
  {
   "sql": "CREATE TABLE schema_w1_144.table_w1_145 (col145_w1_146 TIMETZ, col145_w1_147 BIT(19), col145_w1_148 BYTES NULL, INDEX (col145_w1_146), INDEX (col145_w1_148 ASC, col145_w1_147) PARTITION BY LIST (col145_w1_148, col145_w1_147) (PARTITION table145_part_0 VALUES IN (('\\xdc955e4812':::BYTES, B'1000001111101111010'), ('\\x350ef467a1':::BYTES, B'0000000000000000000'), ('\\x8cd02e':::BYTES, B'0110111000001000011'), ('\\x':::BYTES, B'1000111101011110000'), ('\\xe29883':::BYTES, B'0001111110000011001'), ('\\x1f97af0552aa85':::BYTES, B'0011001100111000000'), ('\\xef3839ff':::BYTES, B'0000000000000000000')), PARTITION table145_part_1 VALUES IN (('\\x0b9dba':::BYTES, B'0011010101100111000'), ('\\x2e7f3de1bdf647':::BYTES, B'0111111100001101010'), ('\\xa1f12b':::BYTES, B'1110011000101011001'), ('\\x66':::BYTES, B'0100110011101111001'), ('\\xddc7907dd739':::BYTES, B'1111011111001011000'), ('\\xd465708e7d':::BYTES, B'0101101110101010111'), ('\\x2667c3':::BYTES, B'1011000011001011101')), PARTITION table145_part_2 VALUES IN (('\\x':::BYTES, B'1010011011101011010'), ('\\x2fe647093a27c4e8':::BYTES, B'1111110000110000111'), ('\\xf5ad83':::BYTES, B'1111011001111000111'), ('\\x':::BYTES, B'0101001110011101011'), ('\\x3696dd00e7':::BYTES, B'0010100101000000101'), ('\\x1a21fec09e5f2d3bde':::BYTES, B'1111101100100011101'), ('\\xc47390f9':::BYTES, B'0111111111111111111')), PARTITION table145_part_3 VALUES IN (('\\x5ab427e84652cdc4':::BYTES, B'0111100111001100010'), ('\\x33063c67':::BYTES, B'0101000010010100011'), ('\\x57d40e182b2af697':::BYTES, B'1110100111111110000'), ('\\xff':::BYTES, B'1000110111010101010'), ('\\x6bf2f6':::BYTES, B'0101110000111101010'), ('\\xb91328efb9710c9b':::BYTES, B'1010000111010001011'), ('\\x109f60d14bd29203b2':::BYTES, B'1001010000111100111')), PARTITION table145_part_4 VALUES IN (('\\xcc':::BYTES, B'1110101000010101000'), ('\\x79d8f503':::BYTES, B'0101111110110111101'), ('\\x2b92ad':::BYTES, B'0111110101000101101'), ('\\x':::BYTES, B'0001110111100010110'), ('\\xf8':::BYTES, B'1000110000011001110'), ('\\xd6db436158466d3c':::BYTES, B'0100110110110100101'), ('\\x2149188d8f4e3a1899':::BYTES, B'1111011010111110110')), PARTITION table145_part_5 VALUES IN (('\\x00':::BYTES, B'1010101100001100110'), ('\\x2fe359bdd61405076d':::BYTES, B'1001010010011100011'), ('\\xbe8eac':::BYTES, B'1101001100110010000'), ('\\xddc5b38fa3d5ff86be':::BYTES, B'1101111100101101011'), ('\\x84ddd0':::BYTES, B'1011101011001011001'), ('\\x09ae':::BYTES, B'1001100011011001101'), ('\\x31be12aaf3d062b3':::BYTES, B'1010101111111101010')), PARTITION table145_part_6 VALUES IN (('\\x0718e988b1':::BYTES, B'1011011000100010010'), ('\\x58':::BYTES, B'1000010011010001011'), ('\\xd229b254acbc':::BYTES, B'1101011011100100101'), ('\\xd4b690f01d':::BYTES, B'0101011010000010100'), ('\\xc4cd8f':::BYTES, B'1110000111111001110'), ('\\x8f96f3f77dee':::BYTES, B'0111010111111100001'), ('\\xca99':::BYTES, B'0110111111001000000')), PARTITION table145_part_7 VALUES IN (('\\xa91de28a3ac15764':::BYTES, B'0110101100010000110'), ('\\xff':::BYTES, B'1011111011101111101'), ('\\xbfe785':::BYTES, B'0000000000000000000'), ('\\x964464ba3bad5e8507':::BYTES, B'1011111110111000110'), ('\\x':::BYTES, B'1100101000010111011'), ('\\xf39a48ebc26f9e':::BYTES, B'1101101101011011000'), ('\\xdd92c46792c102':::BYTES, B'0100101110000111001'))), INDEX (col145_w1_147, col145_w1_146 ASC, col145_w1_148), INDEX (col145_w1_147 ASC, col145_w1_148), UNIQUE (col145_w1_146 DESC), INDEX (col145_w1_146 DESC) NOT VISIBLE)",
   "expectedExecErr": "3F000"
  }
 ],
 "expectedExecErrors": "3F000",
 "expectedCommitErrors": "",
 "message": "ROLLBACK; Successfully got expected execution error.: ERROR: cannot create \"schema_w1_144.table_w1_145\" because the target database or schema does not exist (SQLSTATE 3F000)"
}
{
 "workerId": 0,
 "clientTimestamp": "22:07:05.576675",
 "ops": [
  "BEGIN",
  {
   "sql": "ALTER TABLE schema_w0_2.table_w0_86 ADD COLUMN col_w0_86_w0_180 public.enum_w0_20 NULL"
  },
  "COMMIT"
 ],
 "expectedExecErrors": "",
 "expectedCommitErrors": "",
 "message": "TXN RETRY ERROR; ERROR: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE - failed preemptive refresh due to encountered recently written committed value /Table/3/1/121/2/1 @1740694033.434426250,2): \"sql txn\" meta={id=b4c5871c key=/Table/3/1/112/2/1 iso=Serializable pri=0.04215967 epo=0 ts=1740694033.697274462,2 min=1740694025.575974704,0 seq=14} lock=true stat=PENDING rts=1740694025.575974704,0 wto=false gul=1740694026.075974704,0 (SQLSTATE 40001)"
}
{
 "workerId": 1,
 "clientTimestamp": "22:07:05.635122",
 "ops": [
  "BEGIN",
  {
   "sql": "CREATE SEQUENCE IF NOT EXISTS schema_w1_102.seq_w1_149 OWNED BY public.table_w0_41.col41_w0_42"
  },
  "COMMIT"
 ],
 "expectedExecErrors": "",
 "expectedCommitErrors": "",
 "message": ""
}
panic: connection has timed out; sessionID=18282f864bae384e0000000000000001 activeQuery={String:COMMIT TRANSACTION Valid:true}

goroutine 125336 [running]:
github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWatchDog).watchLoop(0x4032842b40, {0x7900cb8, 0x400f3aab90})
	pkg/workload/schemachange/watch_dog.go:113 +0x210
created by github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWatchDog).Start in goroutine 11822
	pkg/workload/schemachange/watch_dog.go:131 +0x114
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #141813 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot T-sql-foundations branch-release-23.2]
- #141696 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot P-2 T-sql-foundations branch-release-23.2.21-rc]
- #141152 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [column &#34;crdb_internal_column_3_name_placeholder&#34; does not exist] [C-test-failure O-robot P-2 T-sql-foundations branch-release-25.1.0-rc]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18959125?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18959125?buildTab=artifacts#/) on master @ [154f10a33364fda219608ab09e20781258abb486](https://github.com/cockroachdb/cockroach/commits/154f10a33364fda219608ab09e20781258abb486):


```
 "clientTimestamp": "16:40:57.0402",
 "ops": [
  "BEGIN",
  {
   "sql": "SELECT t1.col62_w0_63 AS col0, t2.col_w0_58_w0_60 AS col1, t1.col62_w0_74 AS col2, t2.crdb_internal_col58_w0_59_shard_16 AS col3, t1.col_w0_62_w1_24 AS col4, t2.col_w0_58_w0_60 AS col5, t1.col62_w0_70 AS col6 FROM public.table_w1_62 AS t0 ,schema_w0_57.table_w0_62 AS t1 ,public.table_w0_58 AS t2 FETCH FIRST 1 ROWS ONLY",
   "expectedExecErr": "42P01",
   "potentialExecErr": "53100,53200,57014"
  }
 ],
 "expectedExecErrors": "42P01",
 "expectedCommitErrors": "",
 "message": "ROLLBACK; Successfully got expected execution error.: ERROR: relation \"public.table_w1_62\" does not exist (SQLSTATE 42P01)"
}
{
 "workerId": 1,
 "clientTimestamp": "16:41:07.109253",
 "ops": [
  "BEGIN",
  {
   "sql": "INSERT INTO public.table_w0_79 (col79_w0_80, col79_w0_81, col79_w0_82) VALUES ('0102000060E610000003000000BA624B142E7E414032987DBAC2B155C04428A82F3ABEE2411A0980A10FF45940FDE5879781254140016457CBF596F341C5A594AEFD4A43C0C91AB5BF20CC4A40013B89F79BE0F8C1':::GEOGRAPHY,3802:::OID,'[{\"aw6Q\": null, \"b\": 0.270005331816681}, 0.33218540432930155, false]':::JSONB),('01010000A0E610000039379B9CB7CE5D406AA7A41EE1CB1FC047CF5286317EDAC1':::GEOGRAPHY,1009:::OID,'0.9811161219830262':::JSONB)",
   "potentialExecErr": "23502,23505,23514"
  },
  "COMMIT"
 ],
 "expectedExecErrors": "",
 "expectedCommitErrors": "",
 "message": ""
}
{
 "workerId": 1,
 "clientTimestamp": "16:41:18.101136",
 "ops": [
  "BEGIN",
  {
   "sql": "SELECT t2.col54_w1_60 AS col0, t2.col54_w1_55 AS col1, t2.col54_w1_59 AS col2, t0.col79_w0_80 AS col3, t0.col79_w0_83 AS col4, t2.col54_w1_60 AS col5, t0.col79_w0_83 AS col6, t2.col54_w1_58 AS col7, t2.col54_w1_55 AS col8, t0.col79_w0_80 AS col9, t2.col54_w1_59 AS col10 FROM public.table_w0_79 AS t0 ,schema_w0_57.table_w1_63 AS t1 ,schema_w1_29.table_w1_54 AS t2 FETCH FIRST 1 ROWS ONLY",
   "expectedExecErr": "42P01",
   "potentialExecErr": "53100,53200,57014"
  }
 ],
 "expectedExecErrors": "42P01",
 "expectedCommitErrors": "",
 "message": "ROLLBACK; Successfully got expected execution error.: ERROR: relation \"schema_w0_57.table_w1_63\" does not exist (SQLSTATE 42P01)"
}
panic: connection has timed out; sessionID=18286ca2c244819b0000000000000001 activeQuery={String:COMMIT TRANSACTION Valid:true}

goroutine 57820 [running]:
github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWatchDog).watchLoop(0x4011330600, {0x790b698, 0x4004572be0})
	pkg/workload/schemachange/watch_dog.go:113 +0x210
created by github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWatchDog).Start in goroutine 11459
	pkg/workload/schemachange/watch_dog.go:131 +0x114
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #142120 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot T-sql-foundations branch-release-25.1.1-rc release-blocker]
- #141813 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot T-sql-foundations branch-release-23.2]
- #141696 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot P-2 T-sql-foundations branch-release-23.2.21-rc]
- #141152 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [column &#34;crdb_internal_column_3_name_placeholder&#34; does not exist] [C-test-failure O-robot P-2 T-sql-foundations branch-release-25.1.0-rc]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

pkg/ccl/testccl/workload/schemachange/schemachange_test.TestWorkload [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/19013591?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/19013591?buildTab=artifacts#/) on master @ [5d1cb778c3cf15dd688539d7d3ce99c31d742f97](https://github.com/cockroachdb/cockroach/commits/5d1cb778c3cf15dd688539d7d3ce99c31d742f97):


```
{
 "workerId": 0,
 "clientTimestamp": "05:46:47.626729",
 "ops": [
  "BEGIN",
  {
   "sql": "SELECT t2.col_w0_70_w1_78 AS col0, t1.col4_w1_20 AS col1, t2.col70_w0_78 AS col2, t0.col5_w0_17 AS col3, t2.col70_w0_77 AS col4, t1.col4_w1_9 AS col5, t1.col4_w1_10 AS col6, t1.col4_w1_12 AS col7, t0.col_w0_5_w0_20 AS col8, t0.col5_w0_10 AS col9, t0.col5_w0_13 AS col10, t1.col4_w1_9 AS col11, t1.col4_w1_15 AS col12, t2.col70_w0_81 AS col13 FROM schema_w1_48.table_w0_5 AS t0 ,public.table_w1_4 AS t1 ,schema_w1_48.table_w0_70 AS t2 FETCH FIRST 1 ROWS ONLY",
   "potentialExecErr": "53100,53200,57014"
  },
  "COMMIT"
 ],
 "expectedExecErrors": "",
 "expectedCommitErrors": "",
 "message": ""
}
{
 "workerId": 0,
 "clientTimestamp": "05:47:07.443716",
 "ops": [
  "BEGIN",
  {
   "sql": "ALTER TABLE schema_w1_48.table_w0_90 ALTER PRIMARY KEY USING COLUMNS (col90_w0_92, col90_w0_93)",
   "potentialExecErr": "42701,42P10"
  },
  "COMMIT"
 ],
 "expectedExecErrors": "",
 "expectedCommitErrors": "",
 "message": ""
}
{
 "workerId": 0,
 "clientTimestamp": "05:47:29.484989",
 "ops": [
  "BEGIN",
  {
   "sql": "ALTER TABLE schema_w1_48.table_w0_5 ALTER COLUMN col5_w0_16 DROP NOT NULL"
  }
 ],
 "expectedExecErrors": "",
 "expectedCommitErrors": "",
 "message": "ROLLBACK; Ignoring declarative schema changer not implemented error.: ERROR: *tree.AlterTable not implemented in the new schema changer (SQLSTATE XXUUU)"
}
panic: connection has timed out; sessionID=182a2021e71e36190000000000000001 activeQuery={String:COMMIT TRANSACTION Valid:true}

goroutine 61597 [running]:
github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWatchDog).watchLoop(0x400c89bec0, {0x792e358, 0x4009911130})
	pkg/workload/schemachange/watch_dog.go:113 +0x210
created by github.com/cockroachdb/cockroach/pkg/workload/schemachange.(*schemaChangeWatchDog).Start in goroutine 14659
	pkg/workload/schemachange/watch_dog.go:131 +0x114
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #142120 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot P-2 T-sql-foundations branch-release-25.1.1-rc]
- #141696 pkg/ccl/testccl/workload/schemachange/schemachange_test: TestWorkload failed [C-test-failure O-robot P-2 T-sql-foundations branch-release-23.2.21-rc]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWorkload.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/conn_executor_exec.go
+++ b/pkg/sql/conn_executor_exec.go
@@ -2643,14 +2643,14 @@ func (ex *connExecutor) rollbackSQLTransaction(
 	ex.extraTxnState.prepStmtsNamespace.closeAllPortals(ctx, &ex.extraTxnState.prepStmtsNamespaceMemAcc)
 	ex.recordDDLTxnTelemetry(true /* failed */)
 
-	// A non-retryable error automatically rolls-back the transaction if there
-	// are no savepoints. In that case, we can skip rolling-back the transaction
-	// here.
-	wasRolledback := false
+	// A non-retryable error automatically rolls-back the transaction if there are
+	// no savepoints; see the state transition logic in conn_fsm.go. In that case,
+	// we can skip rolling-back the transaction here.
+	isKVTxnOpen := true
 	if _, isAbortedTxn := ex.machine.CurState().(stateAborted); isAbortedTxn {
-		wasRolledback = !ex.state.mu.hasSavepoints
+		isKVTxnOpen = ex.state.mu.txn.IsOpen()
 	}
-	if !wasRolledback {
+	if isKVTxnOpen {
 		if err := ex.state.mu.txn.Rollback(ctx); err != nil {
 			log.Warningf(ctx, "txn rollback failed: %s", err)
 		}

